// graphql/language/schema/kinds.js
/**
 *  Copyright (c) 2015, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 */

// Schema

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var SCHEMA_DOCUMENT = 'SchemaDocument';
exports.SCHEMA_DOCUMENT = SCHEMA_DOCUMENT;
var TYPE_DEFINITION = 'TypeDefinition';
exports.TYPE_DEFINITION = TYPE_DEFINITION;
var FIELD_DEFINITION = 'FieldDefinition';
exports.FIELD_DEFINITION = FIELD_DEFINITION;
var INPUT_VALUE_DEFINITION = 'InputValueDefinition';
exports.INPUT_VALUE_DEFINITION = INPUT_VALUE_DEFINITION;
var INTERFACE_DEFINITION = 'InterfaceDefinition';
exports.INTERFACE_DEFINITION = INTERFACE_DEFINITION;
var UNION_DEFINITION = 'UnionDefinition';
exports.UNION_DEFINITION = UNION_DEFINITION;
var SCALAR_DEFINITION = 'ScalarDefinition';
exports.SCALAR_DEFINITION = SCALAR_DEFINITION;
var ENUM_DEFINITION = 'EnumDefinition';
exports.ENUM_DEFINITION = ENUM_DEFINITION;
var ENUM_VALUE_DEFINITION = 'EnumValueDefinition';
exports.ENUM_VALUE_DEFINITION = ENUM_VALUE_DEFINITION;
var INPUT_OBJECT_DEFINITION = 'InputObjectDefinition';
exports.INPUT_OBJECT_DEFINITION = INPUT_OBJECT_DEFINITION;
